[expo-observe] add expo-router integration docs (#46025)
expo/expo: docs/pages/eas/observe/integrations/expo-router.mdx

---
title: Expo Router integration
sidebar_title: Expo Router
description: Track per-route render and interactive timings by enabling the Expo Router integration for Expo Observe.
---

import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Step } from '~/ui/components/Step';

Expo Observe ships an opt-in integration for [Expo Router](/router/introduction/) that collects per-route metrics tagged with the route pattern (for example, `/(tabs)/sessions/[sessionId]`). This lets you compare navigation performance by route in the dashboard instead of looking only at app-wide aggregates.

## Prerequisites

<Prerequisites>
  <Requirement title="Expo SDK 56 or later">
    The Expo Router integration is available on SDK 56 and later. On earlier SDKs, `expo-observe`
    still tracks app-wide metrics, but per-route navigation events are not emitted.
  </Requirement>
  <Requirement title="An app already using Expo Observe">
    Follow [Get started](/eas/observe/get-started/) to install `expo-observe` and create your first
    build.
  </Requirement>

>   <Requirement title="`expo-router` installed in the app">

Reviewer: I noticed backticks don't render in titles.

```suggestion
  <Requirement title="Expo Router installed in the app">
```

<img width="339" height="109" alt="Image" src="https://github.com/user-attachments/assets/9592a46d-d418-418d-a495-ab1ecd95ee72" />

    The integration depends on `expo-router` at runtime. If the package is not installed, the
    integration becomes a silent no-op.
  </Requirement>
</Prerequisites>

<Step label="1">

## Enable the integration

Call `ExpoObserve.configure()` with the `expo-router` integration flag at module scope, before any screen mounts:

> ```tsx app/_layout.tsx

Reviewer 2: ````suggestion
```tsx src/app/_layout.tsx
````

Because we add examples based on the default template.

import ExpoObserve from 'expo-observe';

ExpoObserve.configure({
  integrations: { 'expo-router': true },
});
```

> > **warning** The integration must be enabled before mount and cannot be toggled at runtime. Calling `configure()` after the app has mounted, or toggling the flag mid-session, throws an error.

Reviewer 2: Let's move this warning at the top of this section.

</Step>

<Step label="2">

## Wrap your root with `ObserveRoot`

`ObserveRoot` includes `AppMetricsRoot` plus the integration context that holds navigation timing state. Replace `AppMetricsRoot.wrap(...)` with `ObserveRoot.wrap(...)`:

> ```tsx app/_layout.tsx

Reviewer 2: ````suggestion
```tsx src/app/_layout.tsx
````

import ExpoObserve, { ObserveRoot } from 'expo-observe';
import { Stack } from 'expo-router';

ExpoObserve.configure({
  integrations: { 'expo-router': true },
});

function RootLayout() {
  return <Stack />;
}

export default ObserveRoot.wrap(RootLayout);
```

> </Step>

Reviewer: I think we should skip this step, it looks like it's more of a guide on how to migrate from 55 -> 56.

<Step label="3">

## Call `useObserve()` in your screens

Use the `useObserve()` hook to get a `markInteractive` that is automatically scoped to the current route. The emitted event is tagged with the screen's route pattern.

> ```tsx app/(tabs)/index.tsx

Reviewer 2: ```suggestion
```tsx src/app/(tabs)/index.tsx
```

| `routeName`   | `string`  | Route pattern, for example `/(tabs)/sessions/[sessionId]`.                 |
| `url`         | `string`  | Resolved pathname for the navigation.                                      |
| `routeParams` | `object`  | Resolved route params (for example, `{ sessionId: 'abc' }`).               |
| `isAppLaunch` | `boolean` | `true` when measured against process start, `false` for later navigations. |

### Per-route warm render (`warm_ttr`)

**What it measures:** Same as `cold_ttr`, but for screens that were already rendered before focus, typically because they were preloaded via [`<Link prefetch />`](/router/basics/navigation/#prefetching) or the user navigated back to them.

**Event params:**

| Param         | Type     | Description                                                  |
| ------------- | -------- | ------------------------------------------------------------ |
| `routeName`   | `string` | Route pattern, for example `/(tabs)/sessions/[sessionId]`.   |
| `url`         | `string` | Resolved pathname for the navigation.                        |
| `routeParams` | `object` | Resolved route params (for example, `{ sessionId: 'abc' }`). |

> ### Per-screen time to interactive (`tti`)

Reviewer: The other 2 say "per route", not "per screen"

```suggestion
### Per-route time to interactive (`tti`)
```

---
title: Expo Router integration
sidebar_title: Expo Router
description: Track per-route render and interactive timings by enabling the Expo Router integration for Expo Observe.
---

import { Prerequisites, Requirement } from '~/ui/components/Prerequisites';
import { Step } from '~/ui/components/Step';

Expo Observe ships an opt-in integration for [Expo Router](/router/introduction/) that collects per-route metrics tagged with the route pattern (for example, `/(tabs)/sessions/[sessionId]`). This lets you compare navigation performance by route in the dashboard instead of looking only at app-wide aggregates.

## Prerequisites

<Prerequisites>
  <Requirement title="Expo SDK 56 or later">
    The Expo Router integration is available on SDK 56 and later. On earlier SDKs, `expo-observe`
    still tracks app-wide metrics, but per-route navigation events are not emitted.
  </Requirement>
  <Requirement title="An app already using Expo Observe">
    Follow [Get started](/eas/observe/get-started/) to install `expo-observe` and create your first
    build.
  </Requirement>
  <Requirement title="`expo-router` installed in the app">

>     The integration soft-requires `expo-router`. If the package is not installed, the integration

Reviewer: ```suggestion
    The integration depends on `expo-router` at runtime. If the package is not installed, the integration becomes a silent no-op.
```

</Step>

<Step label="2">

## Wrap your root with `ObserveRoot`

`ObserveRoot` includes `AppMetricsRoot` plus the integration context that holds navigation timing state. Replace `AppMetricsRoot.wrap(...)` with `ObserveRoot.wrap(...)`:

```tsx app/_layout.tsx
import ExpoObserve, { ObserveRoot } from 'expo-observe';
import { Stack } from 'expo-router';

ExpoObserve.configure({
  integrations: { 'expo-router': true },
});

function RootLayout() {
  return <Stack />;
}

export default ObserveRoot.wrap(RootLayout);
```

</Step>

<Step label="3">

## Call `useObserve()` in your screens

Use the `useObserve()` hook to get a `markInteractive` that is automatically scoped to the current route. The emitted event is tagged with the screen's route pattern.

```tsx app/(tabs)/index.tsx
import { useObserve } from 'expo-observe';
import { useEffect } from 'react';

export default function Home() {
  const { markInteractive } = useObserve();

  useEffect(() => {
    markInteractive();
  }, [markInteractive]);

>   return /* ... */;

Reviewer: ```suggestion
  return (/* your screen content */);
```

}
```

> **info** If the integration is disabled or `expo-router` is not installed, `useObserve()` falls back to the global `AppMetrics.markInteractive`. You can leave the hook in place regardless of integration state.

</Step>

## Metrics

> ### Per-route time to first render (`cold_ttr`)

Reviewer: ```suggestion
### Per-route first render (`cold_ttr`)
```

> **What it measures:** Time from when a navigation action is dispatched - for example link click - to when the destination screen first becomes focused. For the very first focus after app launch, the measurement is taken from when the js bundle is loaded, and the event includes `isAppLaunch: true`.

Reviewer: ```suggestion
**What it measures:** Time from when a navigation action is dispatched (for example, a link click) to when the destination screen first becomes focused. For the very first focus after app launch, the measurement is taken from when the JS bundle is loaded, and the event includes `isAppLaunch: true`.
```

> Emitted at most once per `(route, screenId)` pair within a session.

Reviewer: ```suggestion
Emitted at most once per screen instance within a session.
```

**Event params:**

| Param         | Type      | Description                                                                |
| ------------- | --------- | -------------------------------------------------------------------------- |
| `routeName`   | `string`  | Route pattern, for example `/(tabs)/sessions/[sessionId]`.                 |
| `url`         | `string`  | Resolved pathname for the navigation.                                      |
| `routeParams` | `object`  | Resolved route params (for example, `{ sessionId: 'abc' }`).               |
| `isAppLaunch` | `boolean` | `true` when measured against process start, `false` for later navigations. |

### Per-route warm render (`warm_ttr`)

> **What it measures:** Same as `cold_ttr`, but for screens that were already rendered before focus — typically because they were preloaded via [`<Link prefetch />`](/router/basics/navigation/#prefetching) or the back navigation happened.

Reviewer: ```suggestion
**What it measures:** Same as `cold_ttr`, but for screens that were already rendered before focus, typically because they were preloaded via [`<Link prefetch />`](/router/basics/navigation/#prefetching) or the user navigated back to them.
```

| `url`         | `string` | Resolved pathname for the navigation.                        |
| `routeParams` | `object` | Resolved route params (for example, `{ sessionId: 'abc' }`). |

### Per-screen time to interactive (`tti`)

**What it measures:** Time from when a navigation action is dispatched to when `markInteractive()` is called on the destination screen. Only the first call per navigation is recorded, so it is safe to call `markInteractive()` multiple times.

**Event params:**

| Param         | Type     | Description                                                          |
| ------------- | -------- | -------------------------------------------------------------------- |
| `routeName`   | `string` | Route pattern, for example `/(tabs)/sessions/[sessionId]`.           |
| `url`         | `string` | Resolved pathname.                                                   |
| `routeParams` | `object` | Resolved route params.                                               |
| `...`         | `any`    | Any custom params passed via `markInteractive({ params: { ... } })`. |

### Notes on `routeName`

> `routeName` is a pattern (`/(tabs)/sessions/[sessionId]), not a resolved URL (`/sessions/abc`). This keeps metrics stable across distinct param values so the dashboard buckets them together. Resolved values are still available on the event via `url`and`routeParams`.

Reviewer: This should be in the "Notes and troubleshooting" section

### Preloaded routes

> Calls to `router.prefetch()` do not count as a user navigation and never seed a `cold_ttr` or `warm_ttr` measurement. The next user-driven navigation to that route emits `warm_ttr` because the screen has already rendered.

Reviewer: This as well I think, I'd add it to the notes section.
